### DISCUSSION TITLE
Pocketbase as a framework production deployment

### DISCUSSION THREAD

**Discussion by tonnax** [pocketbase/pocketbase#4444] [Q&A] [answered]
Good day!

I've successfully deployed my local Pocketbase backend (Go framework) to a VPS service that I intended to use for production. After deployment, I discovered several things:

1. Upon accessing the Admin UI in production, all my local records (including Users and Logs) were also uploaded to the production version.
2. My OAuth and S3 bucket configurations from the local setup remained attached in the production version.
3. All local backup files were present in production.
4. My next plan is to manually remove all the records from every collection to start fresh and set up a new OAuth URL for the production environment, meaning I will have two URLs (local and production).

Questions regarding the above findings:

1. The next time I want to deploy a **new/updated** Pocketbase Go framework backend feature/logic from my local environment to my VPS production, what should I do? Should I just replace the old main.go file with the new main.go file (assuming all my backend logic and new logic are contained within the main.go file), or do I need to replace all the files, including the Pocketbase executable file? (I doubt the latter option.) Let me know if you have better suggestions. _**Again, my goal is to prevent local data from being uploaded to my production environment and replacing my production data every time I deploy a new backend feature.**_

2. Do you have any tips or tricks to make future deployments as smooth as possible? Since I have 2 URLs (local and production), I've set up Dart/Client logic to target them based on the build environment type.

Kindly assist, and thank you to the Pocketbase community!

**Comment by ganigeorgiev** [comment]
Your local data (db, storage, settings, etc.) is stored in the `pb_data` directory and has nothing to do with your Go files (_unless you have data seeders written in Go that executes on startup_).

Don't copy your local `pb_data` to your production environment if you don't want to transfer the data.

The only thing that you usually have to deploy on your server when using PocketBase as Go framework is the resulting binary (_if you are extending PocketBase with JS then you'll need to copy also `pb_migrations` and `pb_hooks` folders_).

For more details, please refer to https://pocketbase.io/docs/going-to-production/

**Reply by ganigeorgiev** [reply]
Additionally, if you want to transfer your local collections changes (eg. adding or removing new fields) check the [Migrations guide](https://pocketbase.io/docs/go-migrations/) (the migrations are regular `.go` files and they will be embedded in the binary).

**Reply by tonnax** [reply]
I see, please kindly help verify one last point:
- Next time when I have new feature code on my Pocketbase main.go, I just need to replace main.go from my local to the one on the server, right? ... or this is not how it works?

Newbie here, really appreciate your help. I've ask GPT4, he told me to `GOOS=linux GOARCH=amd64 go build -o yourAppName` hehe which I don't think is necessary

**Reply by ganigeorgiev** [reply]
> Next time when I have new feature code on my Pocketbase main.go, I just need to replace main.go from my local to the one on the server, right? ... or this is not how it works?

I'm not sure that fully understand your deployment setup but again you usually don't have to transfer any `.go` files - the only thing that you need to deploy is the single resulting binary file from the `go build` command (see also https://pocketbase.io/docs/go-overview/).

If you need further help, please explain in details the steps and commands you've executed to deploy your application.

**Reply by tonnax** [reply]
Ok. For context, I'm building Flutter mobile app and am utilizing both Dart SDK for quick query on client and Pocketbase Go framework for more complex app logic.

For the deployment, I'm fully following https://pocketbase.io/docs/going-to-production/ steps here. I use rysnc to move all the files from my Pocketbase Go framework folder to my VPS server, I created the systemd as the docs suggested and my pocketbase now is running good on a specified port. See my files structure here:
![image](https://github.com/pocketbase/pocketbase/assets/140943390/fccb5781-ebd4-4fb3-bf5c-6f2808961c1a)

I never use `go build` command as you mention before even on my local, I'm always using `go run main.go serve` and it works ok including all the app logic I've written on `main.go`, see what I mean by app logic on `main.go` here: (this is what I want to update on my next feature deployment, adding new endpoints etc)
<img width="629" alt="image" src="https://github.com/pocketbase/pocketbase/assets/140943390/9bae15e8-ee2f-447b-a9c6-7e60e469a6b7">

do you mind explain how `single resulting binary file` works regarding the deployment? 
does deployment for using Pocketbase-Go-framework different from just using Pocketbase as SDK?

**Reply by ganigeorgiev** [reply, marked as answer]
If the `pocketbase` file in the above screenshot is the prebuilt executable then you don't need it.

**When you use PocketBase as Go framework you create your own executable/binary.**

This is no different than building and deploying any other Go application.

> do you mind explain how single resulting binary file works regarding the deployment?

I'm sorry, I don't know how and I don't have the patience to guide you further.

My recommendations are:
- Don't blindly trust an AI answers if you don't understand what is outputting.
- Go through a [Tour of Go](https://go.dev/tour/welcome/1).
- Read the PocketBase Go guides, starting from https://pocketbase.io/docs/go-overview/.
- The only thing you need to deploy when using PocketBase as Go framework is the resulting executable/binary from the `go build` command.

**Reply by tonnax** [reply]
Gotcha! I'll study more on how Go application works, you are the best Gani! 🔥

**Reply by tonnax** [reply]
@ganigeorgiev I've successfully generate and serve the executable, works like a charm. 

One question, I'm quite new with the migrations part, my prod Pocketbase admin only have Users collection and not sure why all of my files inside `pb_migrations` are `.js` files even though I'm extending with Go. So I figure out an quick visual way to export my **local collection schema** to my **production Pocketbase** via Settings. 
![image](https://github.com/pocketbase/pocketbase/assets/140943390/2b8c7532-3722-4869-aa18-61b6832b6c1d)

Do you think this is a hassle-free reliable way to migrate fields schema changes from local to production? ✌️

**Reply by ganigeorgiev** [reply]
> not sure why all of my files inside pb_migrations are .js files 

Because most likely you were using the prebuilt executable when the schema changes were applied.

For the Go migrations please read https://pocketbase.io/docs/go-migrations/.
Once the migrate command is enabled you should be able to generate [Collections snapshot](https://pocketbase.io/docs/go-migrations/#collections-snapshot).
